[#152] 카테고리: 카테고리 기능 구현

jwonyLee opened the pull request:
## 개요

- 카테고리와 관련된 기능 구현

## 동작 화면

<img src="https://user-images.githubusercontent.com/15073405/139523878-a126113a-6a2e-42d7-ba35-4846e731405c.gif" width="50%" />

## 작업 사항

- 카테고리 화면에서 텍스트필드에 입력한 내용을 Core Data에 저장한다.
  - `CategoryRepository`를 이용해 데이터를 저장했습니다.
  - `UITextFieldDelegate.textFieldShouldReturn(_:)`를 이용해 키보드 완료 버튼을 탭했을 때, 데이터를 저장합니다.
    - 데이터가 성공적으로 저장되었다면, `textField`를 초기화하고 키보드를 내립니다. (협의된 부분)
- 카테고리 화면에서 네비게이션 우측 상단 확인 버튼을 눌렀을 때 선택한 카테고리 목록을 `WriteViewModel`과 `WriteCategoryTableCell`에 전달한다.
  - 다른 화면으로 데이터를 전달하기 좋은 다른 방법이 생각나지 않아서 `NotificationCenter`를 사용해 데이터를 전달했습니다.
  - `[IndexPath]` → `WriteViewModel`
    - 현재 `WriteViewModel`에서 `[IndexPath]`를 갖고 있어도 사용하는 곳이 없지만, 이후에 글을 저장할 때 사용할 계획입니다.
  - `[(IndexPath, Category)]` → `WriteCategoryTableCell`
    - `IndexPath`를 같이 전달하는 이유는 글쓰기 화면에서 선택된 카테고리를 삭제할 때 사용하기 위함입니다.
- Core Data에 저장된 카테고리 데이터를 표시한다.
  - `CategoryViewController`에서 전달받은 `[(IndexPath, Category)]`가 있다면 `categoryLabel`을 숨기고 `collectionView`를 표시합니다. 
  - 데이터가 없다면 `categoryLabel`을 표시하고 `collectionView`를 숨깁니다.
- 글쓰기 화면에서 선택된 카테고리를 삭제할 수 있다.
  - `CategoryViewController`에서 `IndexPath`를 같이 전달 받은 이유인데, 카테고리 우측 버튼을 탭하면 `NotificationCenter`를 이용해 해당 `IndexPath`를 전달합니다. `WriteViewModel`에서는 이를 감지해 해당 `IndexPath`를 찾고, 삭제합니다.

### Linked Issue

close #152

jwonyLee committed:
[#152] feat: Core Data와 바인딩 - 카테고리 표시

jwonyLee committed:
[#152] feat: Core Data에 카테고리 저장

jwonyLee committed:
[#152] feat: 카테고리 화면에서 선택한 카테고리들을 글쓰기 화면으로 전달

jwonyLee committed:
[#152] feat: 선택한 카테고리를 WriteCategoryTableCell에 전달, 표시

jwonyLee committed:
[#152] feat: 글쓰기 화면에서 선택된 카테고리 삭제 기능 구현

gustn3965 commented on ThingLog/View/TableVeiwCell/WriteCategoryTableCell.swift:
```diff
@@ -114,7 +132,21 @@ final class WriteCategoryTableCell: UITableViewCell {
+    /// categories 값이 비어있는 경우 categoryLabel을 보여주고, collecionView를 숨긴다.
+    /// categories 값이 있는 경우 categoryLabel을 숨기고 collectionView를 보여주고, collectionView.reloadData()를 호출한다.
+    private func updateViewByCategories() {
+        if categories.isEmpty {
+            categoryLabel.isHidden = false
+            collectionView.isHidden = true
+        } else {
+            categoryLabel.isHidden = true
+            collectionView.isHidden = false
+            collectionView.reloadData()
```
이렇게 세줄로 표현하는 건 어떠신가요? 
```swift
categoryLabel.isHidden = !categories.isEmpty 
collectionView.isHidden = categories.isEmpty
collecitonView.reloadData() 
```

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -81,8 +83,16 @@ final class CategoryViewController: UIViewController {
-                self?.coordinator?.back()
+                guard let self = self else { return }
+                // 선택한 카테고리 IndexPath를 WriteViewModel에게 전달한다.
+                NotificationCenter.default.post(name: .passToSelectedCategoryIndexPaths,
+                                                object: nil,
+                                                userInfo: [Notification.Name.passToSelectedCategoryIndexPaths: self.selectedCategoryIndexPaths])
+                // 선택한 카테고리 IndexPath와 Category를 WriteCategoryTableCell에게 전달한다.
+                NotificationCenter.default.post(name: .passToSelectedIndexPathsWithCategory,
+                                                object: nil,
+                                                userInfo: [Notification.Name.passToSelectedIndexPathsWithCategory: self.selectedIndexPathWithCategory()])
```
`WriteViewModel`에 `indexPath`를 사용해서 글을 저장할 때 사용할 계획이라고 했는데, 제가 이해한게 맞다면, `WriteViewController`에서 `Category fetchReuslts Controller`를 이용하여, 받아온 `indexPath`로 `CategoryEntity`들을 참조하여 같이 저장하려는 건가용? 
* 사실` Category Entity`는 그렇게 큰 데이터도 아니고, `CategoryEntity`를 `toModel()`를 이용해서 `Category`로 만들고, `toEntity()`를 이용해서 글을 저장하면 되지 않을까요? 이게 지원님이 작성하신 CoreData의 목적과도 맞다고 생각하고, 또, IndexPath보다 더 직관적인 것 같아서요.  편하게 의견 주세요. 

위의 가정만 맞다면 `WriteCategoryTableCell`에도 `indexPath`까지는 전달할 필요는 없을 것 같아요

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -114,6 +124,16 @@ final class CategoryViewController: UIViewController {
     }
+
+    /// 선택한 카테고리를 IndexPath와 함께 Category 객체로 변환해서 반환한다.
+    private func selectedIndexPathWithCategory() -> [(IndexPath, Category)] {
+        var categories: [(IndexPath, Category)] = []
+        selectedCategoryIndexPaths.forEach { indexPath in
+            let categoryEntity: CategoryEntity = repository.fetchedResultsController.object(at: indexPath)
+            categories.append((indexPath, categoryEntity.toModel()))
+        }
+        return categories
```
지금 `Set`자료구조에서 꺼내와서 순차적으로 배열에 추가하기 때문에, `WriteCategoryTableCell`에서 같은 카테고리를 매번 선택해도 늘 다른 순서로 표시하고 있습니다. 이게 의도한게 맞는지요?  저는 늘 순서가 없는 것 보다 카테고리 선택화면처럼 늘 정렬되어있으면 좋을 것 같은데, 이건 병연님이랑 이야기를 나누셔야 할 것 같습니다.

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -172,30 +194,64 @@ extension CategoryViewController: UITableViewDelegate {
+            return false
+        }
+
+        repository.create(Category(title: name)) { result in
+            switch result {
+            case .success:
+                textField.text = ""
+                textField.sendActions(for: .valueChanged)
+                textField.resignFirstResponder()
+            case .failure(let error):
+                fatalError(error.localizedDescription)
+            }
+        }
+
+        return true
+    }
```
`create` 메소드가 내부적으로 `perform`을 사용하고 있기 때문에,  메인컨텍스트라도 비동기적으로 이루어지니 성공과 실패가 나오기전에도 `return true`가 먼저 호출될 수 있다라고 생각합니다.

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -32,8 +34,43 @@ final class WriteViewModel {
+
+extension WriteViewModel {
+    /// `CategoryViewController`에서 전달받은 데이터를 `selectedCategoryIndexPaths`에 저장한다.
+    private func bindPassToSelectedCategoryIndexPaths() {
```
메소드명이 구체적이여서 좋은데요! 다만 메소드명 자체만 보면 이것이 Notification에서 오는 걸 옵저빙한다는 것인지는 모르기 때문에 Notificatioin도 네이밍에 추가하는 건 어떠신가요?

gustn3965 reviewed (commented):
고생하셨습니다!👍 

작업사항에 최대한 작성해주셔서 코드 읽는데 있어서 굉장히 좋았습니다. 감사합니다.

코멘트들 한번 확인해주세요 ~

동작화면은 대체로 괜찮은 것 같습니다 ~ 

다만 몇가지 있는데요, 

1. 공백만 넣은 경우에도 추가된다는 점. 
2. 현재는 카테고리의 글자수 제한이 없는 것 같고, 의도하신건지는 모르겠지만 화면처럼 글자가 짤리게 됩니다. 
![image](https://user-images.githubusercontent.com/48749182/139525963-bff667ad-b4d3-4896-8d37-c7c8fe276d01.png)

jwonyLee committed:
[#152] ref: 피드백 반영
- updateViewByCategories 코드 수정
- IndexPath를 전달하지 않고 Category 객체만 전달하게 변경
- selectedIndexPathWithCategory -> selectedCategory 로 변경, 정렬해서 반환
- CategoryRepository.create perform -> performAndWait으로 변경
- WriteViewModel.bind*** 메서드 이름 변경

jwonyLee commented:
피드백 반영해서 수정했습니다!

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -49,6 +49,7 @@ final class CategoryViewController: UIViewController {
     private let leadingTrailingConstant: CGFloat = 18.0
     private let topBottomConstant: CGFloat = 12.0
     private var selectedCategoryIndexPaths: Set<IndexPath> = []
+    private let textFieldMaxLength: Int = 21
```
최대 20자인데 21자 맞나요???

jwonyLee replied:
이 부분은 병연님께 따로 말씀 드렸었는데, 20자로 구성하면 마지막 글자에 받침을 넣을 수 없는 현상이 생겨서 부득이하게 21자로 설정했습니다. 그리고 병연님이 요청하신대로 사용자에게 안내하는 문구에는 20자로 표시했습니당

gustn3965 replied:
오 그런 이슈가 있었군요 네 알겠습니다

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -125,14 +122,14 @@ final class CategoryViewController: UIViewController {
-    private func selectedIndexPathWithCategory() -> [(IndexPath, Category)] {
-        var categories: [(IndexPath, Category)] = []
+    /// 선택한 카테고리를 Category 객체로 변환해서 오름차순으로 정렬 후 반환한다.
+    private func selectedCategory() -> [Category] {
```
메소드명이 프로퍼티 이름같다는 느낌이 강하기도 하고, 메서드명은 최대한 동사로 구성하는게 좋다는 것 같고, 그래서 convert나 changeTo로 네이밍을 추가하는 건 어떨까요,

gustn3965 reviewed (approved):
코멘트 확인하시고 수정이 필요한 경우는 추가 커밋하여 머지하시면 될 것 같아요. 고생하셨습니다.

jwonyLee committed:
[#152] ref: 메서드 이름 변경
selectedCategory() -> convertSelectedIndexPathToCategory()
